<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from AllegroHand_Left.urdf.xacro    | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="AllegroHand_Left" xmlns:xacro="http://www.ros.org/wiki/xacro">
  <!-- ======================== BASE PARAMS ========================= -->
  <!-- ======================== FINGER PARAMS ======================== -->
  <!-- full height from joint to tip. when used,
the radius of the finger tip sphere will be subtracted
and one fixed link will be added for the tip. -->
  <!-- ========================= THUMB PARAMS ========================= -->
  <!-- ========================= LIMITS ========================= -->
  <!-- ============================================================================= -->
  <!-- ============================================================================= -->
  <!-- ============================================================================= -->
  <!-- BASE -->
  <link name="base_link">
    <visual>
      <geometry>
        <!--     drawing  estimate  estimate  -->
        <box size="0.0408 0.1130 0.095"/>
      </geometry>
      <origin rpy="0 0 0" xyz="-0.009300 0 -0.0475"/>
      <material name="blue">
        <color rgba="0 0 .8 1"/>
      </material>
    </visual>
  </link>
  <!-- ============================================================================= -->
  <!-- FINGERS -->
  <!--  LEFT HAND due to which finger is number 0 -->
  <!-- ALSO POSITIVE DIRECTIONS MAY HAVE TO CHANGE -->
  <link name="link_0.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.0164"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.0082"/>
      <material name="black">
        <color rgba=".2 .2 .2 1"/>
      </material>
    </visual>
  </link>
  <joint name="joint_0.0" type="revolute">
    <axis xyz="0 0 1"/>
    <limit effort="0" lower="-0.47" upper="0.47" velocity="0"/>
    <parent link="base_link"/>
    <child link="link_0.0"/>
    <origin rpy="0.08726646255 0 0" xyz="0 -0.044981 -0.002298"/>
  </joint>
  <link name="link_1.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.054"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.027"/>
      <material name="blue">
      </material>
    </visual>
  </link>
  <joint name="joint_1.0" type="revolute">
    <limit effort="0" lower="-0.196" upper="1.61" velocity="0"/>
    <axis xyz="0 1 0"/>
    <parent link="link_0.0"/>
    <child link="link_1.0"/>
    <origin xyz="0 0 0.0164"/>
  </joint>
  <link name="link_2.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.0384"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.0192"/>
      <material name="black">
      </material>
    </visual>
  </link>
  <joint name="joint_2.0" type="revolute">
    <axis xyz="0 1 0"/>
    <limit effort="0" lower="-0.174" upper="1.709" velocity="0"/>
    <parent link="link_1.0"/>
    <child link="link_2.0"/>
    <origin xyz="0 0 0.054"/>
  </joint>
  <link name="link_3.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.0237"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.01185"/>
      <material name="blue">
      </material>
    </visual>
  </link>
  <joint name="joint_3.0" type="revolute">
    <axis xyz="0 1 0"/>
    <limit effort="0" lower="-0.227" upper="1.618" velocity="0"/>
    <parent link="link_2.0"/>
    <child link="link_3.0"/>
    <origin xyz="0 0 0.0384"/>
  </joint>
  <link name="link_3.0_tip">
    <visual>
      <geometry>
        <sphere radius="0.015"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.0237"/>
      <material name="white">
        <color rgba=".9 .9 .9 1"/>
      </material>
    </visual>
  </link>
  <joint name="joint_3.0_tip" type="fixed">
    <parent link="link_3.0"/>
    <child link="link_3.0_tip"/>
    <origin xyz="0 0 0"/>
  </joint>
  <link name="link_4.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.0164"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.0082"/>
      <material name="black">
        <color rgba=".2 .2 .2 1"/>
      </material>
    </visual>
  </link>
  <joint name="joint_4.0" type="revolute">
    <axis xyz="0 0 1"/>
    <limit effort="0" lower="-0.47" upper="0.47" velocity="0"/>
    <parent link="base_link"/>
    <child link="link_4.0"/>
    <origin rpy="0.0 0 0" xyz="0 0 0"/>
  </joint>
  <link name="link_5.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.054"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.027"/>
      <material name="blue">
      </material>
    </visual>
  </link>
  <joint name="joint_5.0" type="revolute">
    <limit effort="0" lower="-0.196" upper="1.61" velocity="0"/>
    <axis xyz="0 1 0"/>
    <parent link="link_4.0"/>
    <child link="link_5.0"/>
    <origin xyz="0 0 0.0164"/>
  </joint>
  <link name="link_6.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.0384"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.0192"/>
      <material name="black">
      </material>
    </visual>
  </link>
  <joint name="joint_6.0" type="revolute">
    <axis xyz="0 1 0"/>
    <limit effort="0" lower="-0.174" upper="1.709" velocity="0"/>
    <parent link="link_5.0"/>
    <child link="link_6.0"/>
    <origin xyz="0 0 0.054"/>
  </joint>
  <link name="link_7.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.0237"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.01185"/>
      <material name="blue">
      </material>
    </visual>
  </link>
  <joint name="joint_7.0" type="revolute">
    <axis xyz="0 1 0"/>
    <limit effort="0" lower="-0.227" upper="1.618" velocity="0"/>
    <parent link="link_6.0"/>
    <child link="link_7.0"/>
    <origin xyz="0 0 0.0384"/>
  </joint>
  <link name="link_7.0_tip">
    <visual>
      <geometry>
        <sphere radius="0.015"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.0237"/>
      <material name="white">
        <color rgba=".9 .9 .9 1"/>
      </material>
    </visual>
  </link>
  <joint name="joint_7.0_tip" type="fixed">
    <parent link="link_7.0"/>
    <child link="link_7.0_tip"/>
    <origin xyz="0 0 0"/>
  </joint>
  <link name="link_8.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.0164"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.0082"/>
      <material name="black">
        <color rgba=".2 .2 .2 1"/>
      </material>
    </visual>
  </link>
  <joint name="joint_8.0" type="revolute">
    <axis xyz="0 0 1"/>
    <limit effort="0" lower="-0.47" upper="0.47" velocity="0"/>
    <parent link="base_link"/>
    <child link="link_8.0"/>
    <origin rpy="-0.08726646255 0 0" xyz="0 0.044981 -0.002298"/>
  </joint>
  <link name="link_9.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.054"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.027"/>
      <material name="blue">
      </material>
    </visual>
  </link>
  <joint name="joint_9.0" type="revolute">
    <limit effort="0" lower="-0.196" upper="1.61" velocity="0"/>
    <axis xyz="0 1 0"/>
    <parent link="link_8.0"/>
    <child link="link_9.0"/>
    <origin xyz="0 0 0.0164"/>
  </joint>
  <link name="link_10.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.0384"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.0192"/>
      <material name="black">
      </material>
    </visual>
  </link>
  <joint name="joint_10.0" type="revolute">
    <axis xyz="0 1 0"/>
    <limit effort="0" lower="-0.174" upper="1.709" velocity="0"/>
    <parent link="link_9.0"/>
    <child link="link_10.0"/>
    <origin xyz="0 0 0.054"/>
  </joint>
  <link name="link_11.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.0237"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.01185"/>
      <material name="blue">
      </material>
    </visual>
  </link>
  <joint name="joint_11.0" type="revolute">
    <axis xyz="0 1 0"/>
    <limit effort="0" lower="-0.227" upper="1.618" velocity="0"/>
    <parent link="link_10.0"/>
    <child link="link_11.0"/>
    <origin xyz="0 0 0.0384"/>
  </joint>
  <link name="link_11.0_tip">
    <visual>
      <geometry>
        <sphere radius="0.015"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.0237"/>
      <material name="white">
        <color rgba=".9 .9 .9 1"/>
      </material>
    </visual>
  </link>
  <joint name="joint_11.0_tip" type="fixed">
    <parent link="link_11.0"/>
    <child link="link_11.0_tip"/>
    <origin xyz="0 0 0"/>
  </joint>
  <link name="link_12.0">
    <visual>
      <geometry>
        <box size="0.0358 0.034 0.0455"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0.0091 -0.009 0.0145"/>
      <material name="black">
        <color rgba=".2 .2 .2 1"/>
      </material>
    </visual>
  </link>
  <joint name="joint_12.0" type="revolute">
    <axis xyz="1 0 0"/>
    <limit effort="0" lower="0.263" upper="1.396" velocity="0"/>
    <parent link="base_link"/>
    <child link="link_12.0"/>
    <origin rpy="0 -1.65806278845 1.5707963259" xyz="-0.0182 -0.0168753 -0.07308"/>
  </joint>
  <link name="link_13.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.0177"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.00885"/>
      <material name="blue">
      </material>
    </visual>
  </link>
  <joint name="joint_13.0" type="revolute">
    <axis xyz="0 0 -1"/>
    <limit effort="0" lower="-0.105" upper="1.163" velocity="0"/>
    <parent link="link_12.0"/>
    <child link="link_13.0"/>
    <origin xyz="0 0 0.0375"/>
  </joint>
  <link name="link_14.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.0514"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.0257"/>
      <material name="black">
      </material>
    </visual>
  </link>
  <joint name="joint_14.0" type="revolute">
    <axis xyz="0 1 0"/>
    <limit effort="0" lower="-0.189" upper="1.644" velocity="0"/>
    <parent link="link_13.0"/>
    <child link="link_14.0"/>
    <origin xyz="0 0 0.0177"/>
  </joint>
  <link name="link_15.0">
    <visual>
      <geometry>
        <box size="0.0196 0.0275 0.0393"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.01965"/>
      <material name="blue">
      </material>
    </visual>
  </link>
  <joint name="joint_15.0" type="revolute">
    <axis xyz="0 1 0"/>
    <limit effort="0" lower="-0.162" upper="1.719" velocity="0"/>
    <parent link="link_14.0"/>
    <child link="link_15.0"/>
    <origin xyz="0 0 0.0514"/>
  </joint>
  <link name="link_15.0_tip">
    <visual>
      <geometry>
        <sphere radius="0.015"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.0393"/>
      <material name="white">
        <color rgba=".9 .9 .9 1"/>
      </material>
    </visual>
  </link>
  <joint name="joint_15.0_tip" type="fixed">
    <parent link="link_15.0"/>
    <child link="link_15.0_tip"/>
    <origin xyz="0 0 0"/>
  </joint>
  <!-- -${95*DEG2RAD}    ${180*DEG2RAD} -->
  <!-- ============================================================================= -->
  <!-- LEFT -->
  <!-- THUMB -->
  <!-- Write code for left and right with proper changes
	(using variables for like comps) and include as 2 macros
	include both and comment unused for right and left -->
  <!-- END LEFT -->
  <!-- ============================================================================= -->
  <!-- ============================================================================= -->
  <!-- ============================================================================= -->
  <!-- THUMB MACRO -->
  <!-- [[END]] THUMB MACRO -->
  <!-- THREE FINGER MACRO -->
  <!-- [[END]] THREE FINGER MACRO -->
</robot>

